### ISSUE TITLE
b:selectMultiMenu returns a wrong String if using selectItems

### ISSUE THREAD

**Issue by JoeBeer** [TheCoder4eu/BootsFaces-OSP#952]
Hi, 
I got a problem that I try to solve for days now and I hope someone can help me. 

Whenever I select multiple rows from a list or set, the string that comes back does not save all selected rows. For example, if I should get 1, 2, 3, 4 back as a string, I get 1, 2, 3 back and if I actually get 1,2,... 10, I get 1, 2, 3, 4, 6,...10 back (the 5 is missing). It varies constantly. 
But if I enter my list or set line by line using selectItem, everything works correctly using itemValue.  

Part of my code:
```xml
<b:selectMultiMenu value="#{registrierenManagedBean.selectedSprachenString}">
      <f:selectItems value="#{registrierenManagedBean.allSprachenIDList}"/>
```

In my case it is about registering a user and selecting languages from a selectMultiMenu, which the user can speak and later edit the selection again. After registration, the user is stored in a database, where it becomes apparent that something is missing. 

I hope I could make my problem clear.

**Comment by chongma** [comment]
How did you specify `allSprachenIDList` in `registrierenManagedBean`? Is it a `List<SelectItem>`?

**Comment by JoeBeer** [comment]
It is currently a `List<Integer>`, had originally tried it with a Sprache object `List<Sprache>` and with a `Set<Sprache>` / `Set<Integer>` . Should I use it with `List<SelectItem>` ?

**Comment by JoeBeer** [comment]
I tried with `<f:selectItems var="sprache" value="#{registrierenManagedBean.allSprachenIDList}" itemLabel="#{sprache.name}" itemValue="#{sprache.id}" />`, but it didn't worked. Had the same problem with the String. 
I tried with `List<SelectetItem>`, but couldn't make it. I'm currently learning Java and couldn't get it to work with `List<SelectetItem>`.  Do you have an example for me how to use it with `List<SelectetItem>`? Thank you.

**Comment by chongma** [comment]
here is an example.  backing bean:
```java
@Named
@ViewScoped
public class RegistrierenManagedBean {
    private List<Sprachen> sprachens;
    private List<SelectItem> allSprachenIDList;
    private List<Integer> selectedSprachen;

    public void onload() {
        //load the sprachens YOUR CODE
       setSprachens(myDAO.selectSprachens());
        initialise();
    }

    private void initialise() {
        setAllSprachenIDList(new ArrayList<>());
        for (Sprachen sprachen : getSprachens()) {
              getAllSprachenIDList.add(new SelectItem(sprachen.getId(), sprachen.getName()));
        }
    }

    private String check() {
        for (Integer sprachen : getSelectedSprachen()) {
            System.out.println("Sprachen: " + sprachen);
        }
        return null;
    }

    ...getters and setters

}
```
and xhtml
```xml
...
<f:metadata>
    <f:viewAction action="#{myBean.onload}" />
</f:metadata>
...
<h:form>
<b:selectMultiMenu value="#{registrierenManagedBean.selectedSprachen}">
    <f:selectItems value="#{registrierenManagedBean.allSprachenIDList}"/>
</b:selectMultiMenu>
<b:commandButton value="save" action="#{registrierenManagedBean.check}">
    <f:ajax execute="@form" render="@form"/>
</b:commandButton>
</h:form>
```

**Comment by JoeBeer** [comment]
I implemented your example, but unfortunately I have the same error again that the string does not save all selected languages. Here are the relevant code digits:

My whole RegistrierenManagedBean.class:

```
@ManagedBean
@RequestScoped
public class RegistrierenManagedBean {

    // =========================== Class Variables ===========================79
    // =============================  Variables  =============================79
    @EJB
    private DAO                             dao;
    private Nutzer                          nutzer;
    private int                             bezirkID;
    private ArrayList<Sprache>              allSprachenList;
    private ArrayList<SelectItem>           selectItemList;
    private List<Sprache>                   selectedSprachenList;
    private String                          selectedSprachenString;
    private ArrayList<Freizeitaktivitaeten> selectedFreizeitaktivitaetenList;
    private String                          selectedFreizeitaktivitaetenString;

    // ============================  Constructors  ===========================79
    public RegistrierenManagedBean() {
        this.nutzer                             = new Nutzer();
        this.selectedSprachenList               = new ArrayList<>();
        this.selectedFreizeitaktivitaetenList   = new ArrayList<>();
    }

    // ===========================  public  Methods  =========================79
    public String register() {
            nutzer.addBezirk(dao.findBezirkByID(bezirkID));
            addSprachenToNutzer();
            addFreizeitaktivitaetenToNutzer();
            dao.merge(nutzer);
            initNutzer();
            return "home";
    }

    public Nutzer getNutzer() {
        return nutzer;
    }

    public void setNutzer(Nutzer nutzer) {
        this.nutzer = nutzer;
    }

    public int getBezirkID() {
        return bezirkID;
    }

    public void setBezirkID(int bezirkID) {
        this.bezirkID = bezirkID;
    }

    public ArrayList<Sprache> getAllSprachenList() {
        allSprachenList = new ArrayList<>();
        allSprachenList = dao.findSprachen();
        return allSprachenList;
    }

    public void setAllSprachenList(ArrayList<Sprache> allSprachenList) {
        this.allSprachenList = allSprachenList;
    }

    public ArrayList<SelectItem> getSelectItemList() {
        selectItemList = new ArrayList<>();
        for (Sprache sprachen : getAllSprachenList()) {
            selectItemList.add(new SelectItem(sprachen.getId(), sprachen.getNameSprache()));
        }
        return selectItemList;
    }

    public void setSelectItemList(ArrayList<SelectItem> selectItemList) {
        this.selectItemList = selectItemList;
    }

    public String getSelectedSprachenString() {
        return selectedSprachenString;
    }

    public void setSelectedSprachenString(String selectedSprachenString) {
        this.selectedSprachenString = selectedSprachenString;
    }

    public String getSelectedFreizeitaktivitaetenString() {
        return selectedFreizeitaktivitaetenString;
    }

    public void setSelectedFreizeitaktivitaetenString(String selectedFreizeitaktivitaetenString) {
        this.selectedFreizeitaktivitaetenString = selectedFreizeitaktivitaetenString;
    }

    // =================  protected/package local  Methods ===================79
    // ===========================  private  Methods  ========================79
    private void addSprachenToNutzer() {
        String[] selectedSprachenArray = selectedSprachenString.split(",");
        for (String aSelectedSprachenArray : selectedSprachenArray) {
            selectedSprachenList.add(dao.findSpracheByID(aSelectedSprachenArray));
        }
        for (Sprache aSelectedSprachenList : selectedSprachenList) {
            nutzer.addSprache(aSelectedSprachenList);
        }
    }

    private void addFreizeitaktivitaetenToNutzer() {
        String[] selectedFreizeitaktivitaetenArray = selectedFreizeitaktivitaetenString.split(",");
        for (String aSelectedFreizeitaktivitaetenArray : selectedFreizeitaktivitaetenArray) {
            selectedFreizeitaktivitaetenList.add(dao.findFreizeitaktivitaetenByID(aSelectedFreizeitaktivitaetenArray));
        }
        for (Freizeitaktivitaeten aSelectedFrezeitaktivitaetenList : selectedFreizeitaktivitaetenList) {
            nutzer.addFreizeitaktivitaeten(aSelectedFrezeitaktivitaetenList);
        }
    }

    private void initNutzer() {
        ELContext elContext = FacesContext.getCurrentInstance().getELContext();
        LoginManagedBean loginManagedBean = (LoginManagedBean) elContext.getELResolver().getValue(elContext, null, "loginManagedBean");
        loginManagedBean.nutzer = dao.findNutzerByMail(nutzer.getMail());
    }

    // ============================  Inner Classes  ==========================79
    // ============================  End of class  ===========================79
}
```

The part of my registrieren.xhtml with the selectMultiMenu: 

```
<b:selectMultiMenu id="ichSpreche" label="Ich spreche:" required="true"
                                       nonSelectedText="Bitte wählen" style="width: 100%"
                                       requiredMessage="Bitte gib mindestens eine Sprache an!"
                                       maxHeight="200" value="#{registrierenManagedBean.selectedSprachenString}"
                                       ondropdownhide="console.log('hide brand');ajax:selectMultiMenuBean.updateBrandMessage()"
                                       update="m_ichSpreche"
                                       process="@this">
                        <f:selectItems var="sprache"
                                       value="#{registrierenManagedBean.selectItemList}"/>
                    </b:selectMultiMenu>
```

When I tried to use a string array as value in the selectMultiMenu, I had an exception and could see the generated string. I had selected the first four entries of my SelectetItemList and the following string could not be passed to the array: 

`java.lang.IllegalArgumentException: Cannot convert 3,1,3,2 of type class java.lang.String to class [Ljava.lang.String;`

Shouldn't the string created by Bootsfaces be `1, 2, 3, 4` ?

**Comment by chongma** [comment]
Although unrelated to the problem in my experience it is bad practice to put code inside getters and setters. I think you need a ViewScoped bean rather than a RequestScoped otherwise it will not be kept alive during the view cycle. Your example is quite complex. You should try something simpler until you get it working the way you want

**Comment by JoeBeer** [comment]
Thank you for your advice and effort. I had already tried as simply as possible and always had the same result. I'm starting to think it's a bug. 
I will keep that with the getters and setters in the future and do it similarly as in your example. :)

**Comment by stephanrauh** [comment]
@chongma @JoeBeer  Historically, getters and setters have been introduced to be able to put code inside them :). If that's forbidden, you could use public attributes as well. Actually, I prefer that, and it's very annoying that JSF, Hibernate, and so on don't support that approach.

Nonetheless, in the case of JSF, you are right. Since I've become aware of the problem, I try to read every value only once. But nobody's perfect, so it's possible a single JSF component calls a getter several times. Putting a database access into such a getter results in a huge performance penalty.

Setters are dangerous, too. By default, each input element calls the setter on every request. 

So it's better to collect all the data you need in a `@PostConstruct` initializer and to store it in bean attributes.

**Comment by chongma** [comment]
@stephanrauh i have to disagree with you about that.  getters and setters are to do with encapsulation.  you might want a property to be read/write, read only or write only which is not possible with a public attribute. it also gives a standard way to access class properties (serializable?). https://www.tutorialspoint.com/java/java_encapsulation.htm.
i know lots of people add code in getters and setters but i stopped doing it as it causes loads of problems

**Comment by stephanrauh** [comment]
Brian Goetz recently gave a nice talk about the topic. Sure, getters and setters are about encapsulation, but why do you want to encapsulate anything? In this particular case, the attributes are explicitly exposed. The idea was to be able to modify the attribute without breaking the API. You can rename it, you can move it to another class, and so on.

Funny thing is nobody does that. Most people who're refactoring their code modify the getters and setters, too. Modern IDEs make this a breeze. So in most programs getters and setter are just useless. Even the JVM optimizes them away. It's a bit like installing security doors in your flat, between the kitchen and the living room.

Read-only and write-only attributes are another story. That's a valid use-case.

Also read https://beyondjava.net/combining-object-oriented-programming-and-functional-programming (and maybe there's a recording of Brian's speech somewhere out there?).

**Comment by stephanrauh** [comment]
@JoeBeer You're JSF code uses a bean that's probably just a copy-and-paste error: `ajax:selectMultiMenuBean.updateBrandMessage()`

**Comment by stephanrauh** [comment]
@JoeBeer I've tried to reproduce your bug. It works reliable on my machine. Maybe - just maybe - the bug is caused by calling the DAO frequently. It might return different objects each time (same data, but different memory location). This, in turn, might confuse the `SelectMultiMenuRenderer`. I suggest implementing a `@PostConstruct` initializer and see if it helps.

You can also copy my version of your code into your project. Is my copy unreliable, too? 

```xml
<?xml version='1.0' encoding='UTF-8' ?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"
	xmlns:h="http://java.sun.com/jsf/html"
	xmlns:f="http://java.sun.com/jsf/core"
	xmlns:b="http://bootsfaces.net/ui" xmlns:p="http://primefaces.org/ui">
<h:head>
	<title>BootsFaces rocks!</title>
	<meta name="author" content="Stephan Rauh"></meta>
</h:head>
<h:body>

<h:form>
		<b:selectMultiMenu id="ichSpreche" label="Ich spreche:" required="true"
	                                       nonSelectedText="Bitte wählen" style="width: 100%"
	                                       requiredMessage="Bitte gib mindestens eine Sprache an!"
	                                       maxHeight="200" value="#{registrierenManagedBean.selectedSprachenString}"
	                                       ondropdownhide="console.log('hide brand');ajax:registrierenManagedBean.register()"
	                                       update="m_ichSpreche"
	                                       process="@this">
	                        <f:selectItems var="sprache"
	                                       value="#{registrierenManagedBean.selectItemList}"/>
     </b:selectMultiMenu>
                    
     <b:inputText label="Ausgabe" value="#{registrierenManagedBean.selectedSprachenString}" id="m_ichSpreche" disabled="true" />
</h:form>
</h:body>
</html>
```

```java
package de.beyondjava.jsf;

import java.util.ArrayList;
import java.util.List;

import javax.faces.bean.ManagedBean;
import javax.faces.bean.RequestScoped;
import javax.faces.model.SelectItem;

@ManagedBean
@RequestScoped
public class RegistrierenManagedBean {
    private ArrayList<Sprache>              allSprachenList;
    private ArrayList<SelectItem>           selectItemList;
    private List<Sprache>                   selectedSprachenList;
    private String                          selectedSprachenString;

    public RegistrierenManagedBean() {
        this.selectedSprachenList               = new ArrayList<>();
    }

    public String register() {
            addSprachenToNutzer();
            return "home";
    }


    public ArrayList<Sprache> getAllSprachenList() {
        allSprachenList = new ArrayList<>();
        allSprachenList = findSprachen();
        return allSprachenList;
    }
    
    private ArrayList<Sprache> findSprachen() {
    		ArrayList<Sprache> l = new ArrayList<>();
    		l.add(new Sprache(1, "Englisch"));
    		l.add(new Sprache(2, "Französisch"));
    		l.add(new Sprache(3, "Italienisch"));
    		return l;
    }

    public void setAllSprachenList(ArrayList<Sprache> allSprachenList) {
        this.allSprachenList = allSprachenList;
    }

    public ArrayList<SelectItem> getSelectItemList() {
        selectItemList = new ArrayList<>();
        for (Sprache sprachen : getAllSprachenList()) {
            selectItemList.add(new SelectItem(sprachen.getId(), sprachen.getNameSprache()));
        }
        return selectItemList;
    }

    public void setSelectItemList(ArrayList<SelectItem> selectItemList) {
        this.selectItemList = selectItemList;
    }

    public String getSelectedSprachenString() {
        return selectedSprachenString;
    }

    public void setSelectedSprachenString(String selectedSprachenString) {
        this.selectedSprachenString = selectedSprachenString;
    }

    private void addSprachenToNutzer() {
        String[] selectedSprachenArray = selectedSprachenString.split(",");
        for (String aSelectedSprachenArray : selectedSprachenArray) {
            System.out.println(aSelectedSprachenArray);
        }
        for (Sprache aSelectedSprachenList : selectedSprachenList) {
            System.out.println("Add Nutzer: " + aSelectedSprachenList);
        }
    }
}
```

**Comment by stephanrauh** [comment]
BTW, you can debug the `SelectMultiMenuRenderer` yourself. The relevant part of the code is the `decode()` method, lines 59-127 (https://github.com/TheCoder4eu/BootsFaces-OSP/blob/05169a3760d717a2e8bf80a855c5e6503f3d730e/src/main/java/net/bootsfaces/component/selectMultiMenu/SelectMultiMenuRenderer.java#L59).

If you're using Maven, just open the class `SelectMultiMenuRenderer` and wait until the IDE has loaded the source code. Now you can debug just like you'd debug your own code.

**Comment by JoeBeer** [comment]
@stephanrauh Thank you a lot. I'm on vacation right now and will be back at home tomorrow. Then I will try your code and your suggestions and report it here. :)

**Comment by JoeBeer** [comment]
@stephanrauh I tried with `@PostConstruct`, but still had the same problem. We made the project public, so if you like you can take a look at it. https://github.com/pascalvon/TandemPartnerWeb its in the branch https://github.com/pascalvon/TandemPartnerWeb/tree/Branch_selectMultiMenu_Issue. The relevant classes are DAO.class, RegistrierenManagedBean.class and the registrieren.xhtml. 

I also have the problem that when I select 3 or 2 languages, the given string contains the number 0 and therefore flies a NoResultException. 
Using `@PostConstruct`. I actually create a list that is only generated once. Still, it doesn't work. I'm really desperate. 
I am very grateful for any help.

**Comment by JoeBeer** [comment]
Another thing that don't work for me is if I fill an ArrayList with a few selectetItem's like this:
```
public ArrayList<SelectItem> getSelectItemList() {
        selectItemList = new ArrayList<>();
        selectItemList.add(new SelectItem(1, "Englisch"));
        selectItemList.add(new SelectItem(2, "Deutsch"));
        selectItemList.add(new SelectItem(3, "Französisch"));
        selectItemList.add(new SelectItem(4, "Hindi"));
        selectItemList.add(new SelectItem(5, "Italienisch"));
        return selectItemList;
    }
```
I select 4 object's but it still only writes the first 3 into the db. :(

**Comment by stephanrauh** [comment]
I know I'm demanding a lot - but I don't have your database. BootsFaces is a leisure-time project. So we can't dedicate as much time to BootsFaces as we'd like to.

Would you mind to provide either a version without database or a version using a H2 database that's initialized automatically when the server starts up?

Another option is to hire me as a consultant. That's not for free, but if you're interested, contact me at stephan.rauh@opitz-consulting.com.

**Comment by JoeBeer** [comment]
Since I'm a student, I can't afford to hire you as a consultant. I will try it with the H2 database and write again when I have implemented it. :)

**Comment by stephanrauh** [comment]
@JoeBeer I've found a commit on your repository that reads "fixed selectMultiMenu". Is it what I suspect it to be: the solution to this ticket?

If it is, this may be worth a closer investigation. Did you solve the bug by replacing ´<f:selectItems>`  by a  long list of hard-wired `<f:selectItem>`?

**Comment by stephanrauh** [comment]
I'm closing this ticket because we can't reproduce the bug. Maybe the bug has been solved by the improvements concerning the converters in BootsFaces 1.3.0.
